File: wallet-mobile/app/lib/api/tokenization_api.dart
// Mirrors wallet-web's api/tokenizationApi.ts: wallet-backend's
// tokenization component (the asset-tokenization application/vetting/
// minting/sale pipeline, and the investor-facing purchase/interest/
// early-exit flows against a live sale). Every mutating on-chain step
// (application fee, crypto purchase, early exit) follows the same
// build -> personal_sign digest -> submit pattern as payments_api.dart/
// swaps_api.dart.
import 'http_client.dart';

class ActionProposal {
  ActionProposal({required this.actionId, required this.digestToSign});
  final int actionId;
  final String digestToSign;

  factory ActionProposal.fromJson(Map<String, dynamic> json) =>
      ActionProposal(actionId: json['actionId'] as int, digestToSign: json['digestToSign'] as String);
}

class PurchaseProposal extends ActionProposal {
  PurchaseProposal({required super.actionId, required super.digestToSign, this.paymentAmount});
  final String? paymentAmount;

  factory PurchaseProposal.fromJson(Map<String, dynamic> json) => PurchaseProposal(
        actionId: json['actionId'] as int,
        digestToSign: json['digestToSign'] as String,
        paymentAmount: json['paymentAmount'] as String?,
      );
}

/// Only the fields this app's UI actually reads/writes - the real backend
/// row carries hundreds of asset-class-specific columns; anything else
/// round-trips through [raw] untouched since this app never overwrites
/// fields it didn't set.
class TokenizedAsset {
  TokenizedAsset({
    required this.id,
    required this.status,
    required this.vettingStatus,
    required this.assetName,
    required this.assetCode,
    required this.assetDescription,
    required this.assetQuoteCurrency,
    required this.pricePerToken,
    required this.raw,
  });

  final int id;
  final String status;
  final bool vettingStatus;
  final String assetName;
  final String assetCode;
  final String assetDescription;
  final String assetQuoteCurrency;
  final String pricePerToken;
  final Map<String, dynamic> raw;

  factory TokenizedAsset.fromJson(Map<String, dynamic> json) => TokenizedAsset(
        id: json['id'] as int,
        status: json['status'] as String? ?? '',
        vettingStatus: json['vettingStatus'] as bool? ?? false,
        assetName: json['assetName'] as String? ?? '',
        assetCode: json['assetCode'] as String? ?? '',
        assetDescription: json['assetDescription'] as String? ?? '',
        assetQuoteCurrency: json['assetQuoteCurrency'] as String? ?? '',
        pricePerToken: json['pricePerToken'] as String? ?? '',
        raw: json,
      );
}

class NewApplicationInput {
  NewApplicationInput({
    required this.assetSector,
    required this.assetSubSector,
    required this.assetType,
    required this.assetName,
    required this.assetCode,
    required this.assetDescription,
    required this.assetCountryLocation,
    required this.assetQuoteCurrency,
    required this.numberOfTokenToBeIssued,
    required this.maxNumberOfTokenAvailableForSale,
    required this.pricePerToken,
    this.assetDecimals = 2,
  });

  final String assetSector;
  final String assetSubSector;
  final String assetType;
  final String assetName;
  final String assetCode;
  final String assetDescription;
  final String assetCountryLocation;
  final String assetQuoteCurrency;
  final String numberOfTokenToBeIssued;
  final String maxNumberOfTokenAvailableForSale;
  final String pricePerToken;
  final int assetDecimals;

  Map<String, dynamic> toJson() => {
        'assetSector': assetSector,
        'assetSubSector': assetSubSector,
        'assetType': assetType,
        'assetName': assetName,
        'assetCode': assetCode,
        'assetDescription': assetDescription,
        'assetCountryLocation': assetCountryLocation,
        'assetQuoteCurrency': assetQuoteCurrency,
        'numberOfTokenToBeIssued': numberOfTokenToBeIssued,
        'maxNumberOfTokenAvailableForSale': maxNumberOfTokenAvailableForSale,
        'pricePerToken': pricePerToken,
        'assetDecimals': assetDecimals,
      };
}

class TokenizedAssetSubscription {
  TokenizedAssetSubscription({
    required this.id,
    required this.tokenizedAssetId,
    required this.quantity,
    required this.paymentAmount,
    required this.paymentAssetSymbol,
    required this.channel,
  });

  final String id;
  final int tokenizedAssetId;
  final String quantity;
  final String paymentAmount;
  final String paymentAssetSymbol;
  final String channel;

  factory TokenizedAssetSubscription.fromJson(Map<String, dynamic> json) => TokenizedAssetSubscription(
        id: json['id'] as String,
        tokenizedAssetId: json['tokenizedAssetId'] as int,
        quantity: json['quantity'] as String? ?? '',
        paymentAmount: json['paymentAmount'] as String? ?? '',
        paymentAssetSymbol: json['paymentAssetSymbol'] as String? ?? '',
        channel: json['channel'] as String? ?? '',
      );
}

class ExpressionOfInterest {
  ExpressionOfInterest({required this.id, required this.tokenizedAssetId, this.amount});
  final int id;
  final int tokenizedAssetId;
  final String? amount;

  factory ExpressionOfInterest.fromJson(Map<String, dynamic> json) => ExpressionOfInterest(
        id: json['id'] as int,
        tokenizedAssetId: json['tokenizedAssetId'] as int,
        amount: json['amount'] as String?,
      );
}

class TokenizedAssetEarlyExit {
  TokenizedAssetEarlyExit({
    required this.id,
    required this.tokenizedAssetId,
    required this.tokenQuantityExited,
    required this.payoutCurrency,
    required this.estimatedPayoutAmount,
  });

  final int id;
  final int tokenizedAssetId;
  final String tokenQuantityExited;
  final String payoutCurrency;
  final String estimatedPayoutAmount;

  factory TokenizedAssetEarlyExit.fromJson(Map<String, dynamic> json) => TokenizedAssetEarlyExit(
        id: json['id'] as int,
        tokenizedAssetId: json['tokenizedAssetId'] as int,
        tokenQuantityExited: json['tokenQuantityExited'] as String? ?? '',
        payoutCurrency: json['payoutCurrency'] as String? ?? '',
        estimatedPayoutAmount: json['estimatedPayoutAmount'] as String? ?? '',
      );
}

class TokenizationApi {
  TokenizationApi(this._client);
  final ApiClient _client;

  Future<TokenizedAsset> submitApplication(String walletAddress, NewApplicationInput input) {
    return _client.request(
      '/v1/tokenization',
      method: 'POST',
      walletAddress: walletAddress,
      body: input.toJson(),
      decode: (json) => TokenizedAsset.fromJson(json as Map<String, dynamic>),
    );
  }

  Future<List<TokenizedAsset>> listMyApplications(String walletAddress) {
    return _client.request(
      '/v1/tokenization/applications',
      walletAddress: walletAddress,
      decode: (json) => (json as List).map((e) => TokenizedAsset.fromJson(e as Map<String, dynamic>)).toList(),
    );
  }

  Future<TokenizedAsset> getApplication(String walletAddress, int assetId) {
    return _client.request(
      '/v1/tokenization/$assetId',
      walletAddress: walletAddress,
      decode: (json) => TokenizedAsset.fromJson(json as Map<String, dynamic>),
    );
  }

  Future<void> deleteApplication(String walletAddress, int assetId) {
    return _client.request(
      '/v1/tokenization/$assetId',
      method: 'DELETE',
      walletAddress: walletAddress,
      decode: (_) {},
    );
  }

  /// Moves DRAFT -> APPLICATION_CONFIRMED and, if a fee is owed, returns a
  /// proposal for it; feePayment is null when no fee applies.
  Future<ActionProposal?> confirmApplication(String walletAddress, int assetId) async {
    final result = await _client.request(
      '/v1/tokenization/$assetId/confirm',
      method: 'PUT',
      walletAddress: walletAddress,
      decode: (json) => json as Map<String, dynamic>,
    );
    final feePayment = result['feePayment'] as Map<String, dynamic>?;
    return feePayment == null ? null : ActionProposal.fromJson(feePayment);
  }

  Future<void> submitApplicationFee(String walletAddress, int assetId, int actionId, String signature) {
    return _client.request(
      '/v1/tokenization/$assetId/confirm/pay-fee',
      method: 'POST',
      walletAddress: walletAddress,
      body: {'actionId': actionId, 'signature': signature},
      decode: (_) {},
    );
  }

  Future<TokenizedAsset> confirmFeePayment(String walletAddress, int assetId) {
    return _client.request(
      '/v1/tokenization/$assetId/fee/confirm',
      method: 'POST',
      walletAddress: walletAddress,
      decode: (json) => TokenizedAsset.fromJson(json as Map<String, dynamic>),
    );
  }

  Future<PurchaseProposal> buildCryptoPurchase(String walletAddress, int assetId, String quantity) {
    return _client.request(
      '/v1/tokenization/$assetId/subscribe',
      method: 'POST',
      walletAddress: walletAddress,
      body: {'quantity': quantity},
      decode: (json) => PurchaseProposal.fromJson(json as Map<String, dynamic>),
    );
  }

  Future<TokenizedAssetSubscription> confirmCryptoPurchase(
    String walletAddress,
    int assetId,
    String quantity,
    int actionId,
    String signature,
  ) {
    return _client.request(
      '/v1/tokenization/$assetId/subscribe/confirm',
      method: 'POST',
      walletAddress: walletAddress,
      body: {'quantity': quantity, 'actionId': actionId, 'signature': signature},
      decode: (json) => TokenizedAssetSubscription.fromJson(json as Map<String, dynamic>),
    );
  }

  Future<List<TokenizedAssetSubscription>> listMySubscriptions(String walletAddress) {
    return _client.request(
      '/v1/tokenization/subscriptions/mine',
      walletAddress: walletAddress,
      decode: (json) => (json as List).map((e) => TokenizedAssetSubscription.fromJson(e as Map<String, dynamic>)).toList(),
    );
  }

  Future<ExpressionOfInterest> expressInterest(String walletAddress, int assetId, {String? amount}) {
    return _client.request(
      '/v1/tokenization/$assetId/interest',
      method: 'POST',
      walletAddress: walletAddress,
      body: {'amount': amount},
      decode: (json) => ExpressionOfInterest.fromJson(json as Map<String, dynamic>),
    );
  }

  Future<List<ExpressionOfInterest>> listMyInterest(String walletAddress) {
    return _client.request(
      '/v1/tokenization/interest/mine',
      walletAddress: walletAddress,
      decode: (json) => (json as List).map((e) => ExpressionOfInterest.fromJson(e as Map<String, dynamic>)).toList(),
    );
  }

  Future<PurchaseProposal> buildEarlyExit(String walletAddress, int assetId, String quantity) {
    return _client.request(
      '/v1/tokenization/$assetId/early-exit',
      method: 'POST',
      walletAddress: walletAddress,
      body: {'quantity': quantity},
      decode: (json) => PurchaseProposal.fromJson(json as Map<String, dynamic>),
    );
  }

  Future<TokenizedAssetEarlyExit> confirmEarlyExit(
    String walletAddress,
    int assetId,
    String quantity,
    int bankId,
    String accountNumber,
    String accountName,
    int actionId,
    String signature,
  ) {
    return _client.request(
      '/v1/tokenization/$assetId/early-exit/confirm',
      method: 'POST',
      walletAddress: walletAddress,
      body: {
        'quantity': quantity,
        'bankId': bankId,
        'accountNumber': accountNumber,
        'accountName': accountName,
        'actionId': actionId,
        'signature': signature,
      },
      decode: (json) => TokenizedAssetEarlyExit.fromJson(json as Map<String, dynamic>),
    );
  }

  Future<List<TokenizedAssetEarlyExit>> listMyEarlyExits(String walletAddress) {
    return _client.request(
      '/v1/tokenization/early-exits/mine',
      walletAddress: walletAddress,
      decode: (json) => (json as List).map((e) => TokenizedAssetEarlyExit.fromJson(e as Map<String, dynamic>)).toList(),
    );
  }
}

File: wallet-web/app/src/api/tokenizationApi.ts
// Client for wallet-backend's tokenization component (PLAN.md §18/19):
// the asset-tokenization application/vetting/minting/sale pipeline, and
// the investor-facing purchase/interest/early-exit flows against a live
// sale. Every mutating on-chain step (application fee, crypto purchase,
// early exit) follows the same build -> personal_sign digest -> submit
// pattern as paymentsApi.ts/swapsApi.ts, per wallet-backend PLAN.md §18's
// fix delegating these to sharedaccess instead of a plain unsigned
// transaction a Safe wallet could never actually sign.
import { apiRequest } from './httpClient';

export interface ActionProposal {
  actionId: number;
  digestToSign: string;
}

export interface PurchaseProposal extends ActionProposal {
  paymentAmount?: string;
}

// Only the fields this app's UI actually reads/writes - the real backend
// row carries hundreds of asset-class-specific columns (PLAN.md §18's own
// note on why a field-for-field form is out of scope here); anything else
// round-trips through the object untouched since this app never
// overwrites fields it didn't set.
export interface TokenizedAsset {
  id: number;
  status: string;
  vettingStatus: boolean;
  assetSector: string;
  assetSubSector: string;
  assetType: string;
  assetName: string;
  assetCode: string;
  assetDescription: string;
  assetCountryLocation: string;
  assetQuoteCurrency: string;
  numberOfTokenToBeIssued: string;
  maxNumberOfTokenAvailableForSale: string;
  pricePerToken: string;
  assetDecimals: number;
  currentNavPerToken?: string;
  earlyExitPenaltyPercent?: string;
  earlyExitFeePercent?: string;
  tokenizationFeeId?: number | null;
  tokenizationApplicationFee?: string;
  tokenizationApplicationFeeAsset?: string;
  saleContractAddress?: string | null;
  issuerContractAddress?: string | null;
  createdAt: string;
  [key: string]: unknown;
}

export interface NewApplicationInput {
  assetSector: string;
  assetSubSector: string;
  assetType: string;
  assetName: string;
  assetCode: string;
  assetDescription: string;
  assetCountryLocation: string;
  assetQuoteCurrency: string;
  numberOfTokenToBeIssued: string;
  maxNumberOfTokenAvailableForSale: string;
  pricePerToken: string;
  assetDecimals: number;
}

export interface TokenizedAssetSubscription {
  id: string;
  tokenizedAssetId: number;
  quantity: string;
  paymentAmount: string;
  paymentAssetSymbol: string;
  txHash: string;
  channel: string;
  createdAt: string;
}

export interface ExpressionOfInterest {
  id: number;
  tokenizedAssetId: number;
  amount: string;
  createdAt: string;
}

export interface TokenizedAssetEarlyExit {
  id: number;
  tokenizedAssetId: number;
  tokenQuantityExited: string;
  payoutCurrency: string;
  payoutPricePerToken: string;
  estimatedPayoutAmount: string;
  burnTxHash: string;
  createdAt: string;
}

// Sector/sub-sector ids are the display strings themselves (e.g.
// "REAL_ESTATE") - the backend's reference tables carry no separate name
// column (PLAN.md §4.9's stakeholder-directory pattern).
export interface ReferenceData {
  sectors: { id: string }[];
  subSectors: { id: string; assetSectorId: string }[];
  currencies: { symbol: string; label: string }[];
  allowedCountries: { countryCode: string }[];
}

export interface TokenizedAssetType {
  id: number;
  assetSubSectorId: string;
  assetType: string;
}

export function getPublicReferenceData(): Promise<ReferenceData> {
  return apiRequest<ReferenceData>('/v1/tokenization/public');
}

export function getFullReferenceData(walletAddress: string): Promise<ReferenceData> {
  return apiRequest<ReferenceData>('/v1/tokenization/reference', { walletAddress });
}

export function getTypesBySubSector(walletAddress: string, subSectorId: string): Promise<TokenizedAssetType[]> {
  return apiRequest<TokenizedAssetType[]>(`/v1/tokenization/types?subSectorId=${encodeURIComponent(subSectorId)}`, {
    walletAddress,
  });
}

export function submitApplication(walletAddress: string, input: NewApplicationInput): Promise<TokenizedAsset> {
  return apiRequest<TokenizedAsset>('/v1/tokenization', { method: 'POST', walletAddress, body: input });
}

export function listMyApplications(walletAddress: string): Promise<TokenizedAsset[]> {
  return apiRequest<TokenizedAsset[]>('/v1/tokenization/applications', { walletAddress });
}

export function getApplication(walletAddress: string, assetId: number): Promise<TokenizedAsset> {
  return apiRequest<TokenizedAsset>(`/v1/tokenization/${assetId}`, { walletAddress });
}

export function deleteApplication(walletAddress: string, assetId: number): Promise<void> {
  return apiRequest<void>(`/v1/tokenization/${assetId}`, { method: 'DELETE', walletAddress });
}

// confirmApplication moves DRAFT -> APPLICATION_CONFIRMED and, if a fee is
// owed, returns a proposal for it; feePayment is null when no fee applies.
export function confirmApplication(
  walletAddress: string,
  assetId: number,
): Promise<{ asset: TokenizedAsset; feePayment: ActionProposal | null }> {
  return apiRequest(`/v1/tokenization/${assetId}/confirm`, { method: 'PUT', walletAddress });
}

export function submitApplicationFee(walletAddress: string, assetId: number, actionId: number, signature: string): Promise<void> {
  return apiRequest<void>(`/v1/tokenization/${assetId}/confirm/pay-fee`, {
    method: 'POST',
    walletAddress,
    body: { actionId, signature },
  });
}

export function confirmFeePayment(walletAddress: string, assetId: number): Promise<TokenizedAsset> {
  return apiRequest<TokenizedAsset>(`/v1/tokenization/${assetId}/fee/confirm`, { method: 'POST', walletAddress });
}

export function buildCryptoPurchase(walletAddress: string, assetId: number, quantity: string): Promise<PurchaseProposal> {
  return apiRequest<PurchaseProposal>(`/v1/tokenization/${assetId}/subscribe`, {
    method: 'POST',
    walletAddress,
    body: { quantity },
  });
}

export function confirmCryptoPurchase(
  walletAddress: string,
  assetId: number,
  quantity: string,
  actionId: number,
  signature: string,
): Promise<TokenizedAssetSubscription> {
  return apiRequest<TokenizedAssetSubscription>(`/v1/tokenization/${assetId}/subscribe/confirm`, {
    method: 'POST',
    walletAddress,
    body: { quantity, actionId, signature },
  });
}

export function listMySubscriptions(walletAddress: string): Promise<TokenizedAssetSubscription[]> {
  return apiRequest<TokenizedAssetSubscription[]>('/v1/tokenization/subscriptions/mine', { walletAddress });
}

export function expressInterest(walletAddress: string, assetId: number, amount?: string): Promise<ExpressionOfInterest> {
  return apiRequest<ExpressionOfInterest>(`/v1/tokenization/${assetId}/interest`, {
    method: 'POST',
    walletAddress,
    body: { amount },
  });
}

export function listMyInterest(walletAddress: string): Promise<ExpressionOfInterest[]> {
  return apiRequest<ExpressionOfInterest[]>('/v1/tokenization/interest/mine', { walletAddress });
}

export function buildEarlyExit(walletAddress: string, assetId: number, quantity: string): Promise<PurchaseProposal> {
  return apiRequest<PurchaseProposal>(`/v1/tokenization/${assetId}/early-exit`, {
    method: 'POST',
    walletAddress,
    body: { quantity },
  });
}

export function confirmEarlyExit(
  walletAddress: string,
  assetId: number,
  quantity: string,
  bankId: number,
  accountNumber: string,
  accountName: string,
  actionId: number,
  signature: string,
): Promise<TokenizedAssetEarlyExit> {
  return apiRequest<TokenizedAssetEarlyExit>(`/v1/tokenization/${assetId}/early-exit/confirm`, {
    method: 'POST',
    walletAddress,
    body: { quantity, bankId, accountNumber, accountName, actionId, signature },
  });
}

export function listMyEarlyExits(walletAddress: string): Promise<TokenizedAssetEarlyExit[]> {
  return apiRequest<TokenizedAssetEarlyExit[]>('/v1/tokenization/early-exits/mine', { walletAddress });
}
